File: 2023/day02/haskell/app/Main.hs
module Main where

import Data.Char (isSpace)
import Data.List (lines)
import Data.List.Split (splitOn)

redLimit :: Int
redLimit = 12

greenLimit :: Int
greenLimit = 13

blueLimit :: Int
blueLimit = 14

newtype Round = Round (Int, Int, Int)
  deriving (Show)

type Game = [Round]

parseLine :: [Char] -> Game
parseLine line = p $ map (map (parseItem . trim) . splitOn ",") $ splitOn ";" lineNoPrefix
  where
    lineNoPrefix = drop 2 $ dropWhile (/= ':') line

    parseItem :: String -> (Int, String)
    parseItem "" = error "Empty item"
    parseItem item = (\(x : xs) -> (read x :: Int, head xs)) $ splitOn " " item

    trim :: String -> String
    trim = f . f
      where
        f = reverse . dropWhile isSpace

    p :: [[(Int, String)]] -> Game
    p = map (foldl q (Round (0, 0, 0)))

    q :: Round -> (Int, String) -> Round
    q (Round (r, g, b)) (n, m)
      | m == "red" = Round (r + n, g, b)
      | m == "green" = Round (r, g + n, b)
      | m == "blue" = Round (r, g, b + n)
      | otherwise = Round (r, g, b)

verify :: Game -> Bool
verify = all possible
  where
    possible (Round (r, g, b)) = r <= redLimit && g <= greenLimit && b <= blueLimit

power :: Game -> Int
power xs = maximum reds * maximum greens * maximum blues
  where
    reds = filter (/= 0) $ map (\(Round (r, _, _)) -> r) xs
    greens = filter (/= 0) $ map (\(Round (_, g, _)) -> g) xs
    blues = filter (/= 0) $ map (\(Round (_, _, b)) -> b) xs

part1 :: [Game] -> Int
part1 games = sum $ map fst $ filter snd $ zip [1 ..] $ map verify games

part2 :: [Game] -> Int
part2 games = sum $ map power games

main :: IO ()
main = do
  input <- lines <$> readFile "../input.txt"

  let games = map parseLine input

  putStrLn $ "Part 1: " ++ show (part1 games)
  putStrLn $ "Part 2: " ++ show (part2 games)

File: 2023/day02/go/src/main.go
package main

import (
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
)

func part1(input string) int {
	sum := 0

	for line_number, line := range strings.Split(input, "\n") {
		rest := strings.Split(line, ": ")[1]
		words := []string{}
		game_is_possible := true
		word_start := 0

		for i := range rest {
			if rest[i] == ';' || rest[i] == ',' || i == len(rest)-1 {
				if i == len(rest)-1 {
					words = append(words, rest[word_start:i+1])
				} else {
					words = append(words, rest[word_start:i])
				}
				word_start = i + 2
			}
		}

		for _, word := range words {
			value, _ := strconv.Atoi(strings.Split(word, " ")[0])
			colour := strings.Split(word, " ")[1]

			switch colour {
			case "red":
				if value > 12 {
					game_is_possible = false
				}
			case "green":
				if value > 13 {
					game_is_possible = false
				}
			case "blue":
				if value > 14 {
					game_is_possible = false
				}
			}
		}

		if game_is_possible {
			sum += line_number + 1
		}
	}
	return sum
}

func part2(input string) int {
	sum := 0

	for _, line := range strings.Split(input, "\n") {
		rest := strings.Split(line, ": ")[1]
		words := []string{}
		word_start := 0

		max_red := 1
		max_green := 1
		max_blue := 1

		for i := range rest {
			if rest[i] == ';' || rest[i] == ',' || i == len(rest)-1 {
				if i == len(rest)-1 {
					words = append(words, rest[word_start:i+1])
				} else {
					words = append(words, rest[word_start:i])
				}
				word_start = i + 2
			}
		}

		for _, word := range words {
			value, _ := strconv.Atoi(strings.Split(word, " ")[0])
			colour := strings.Split(word, " ")[1]

			switch colour {
			case "red":
				max_red = int(math.Max(float64(max_red), float64(value)))
			case "green":
				max_green = int(math.Max(float64(max_green), float64(value)))
			case "blue":
				max_blue = int(math.Max(float64(max_blue), float64(value)))
			}
		}
		sum += max_red * max_green * max_blue
	}
	return sum

}

func main() {
	data, err := os.ReadFile("../input.txt")
	if err != nil {
		panic(err)
	}
	input := string(data)

	fmt.Println("Part 1:", part1(input))
	fmt.Println("Part 2:", part2(input))
}
